File: ruby/lib/redis/active_expire.rb
module Redis
  class ActiveExpire
    CYCLE_INTERVAL = 5.freeze
    TIME_BUDGET = 0.25 * CYCLE_INTERVAL
    
    def self.run(expire_keys:, store:, mutex:)
      Thread.new do
        loop do
          start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
          aggressive_loop_done = false

          until aggressive_loop_done      
            keys_to_delete = mutex.synchronize do
              break [] if expire_keys.empty?

              current_time_ms = Time.now.to_i * 1000
              # pick up random keys and check if they need to expire
              sampled_pairs = expire_keys.to_a.sample(20).to_h
              expired_sample = sampled_pairs.select { |_, v| v < current_time_ms }

              # check the expired key from sample hash
              # is greater than 25% threshold
              expired_key_ratio_above_threshold = (expired_sample.size / sampled_pairs.size) > 0.25
              if expired_key_ratio_above_threshold
                aggressive_loop_done = false
              else
                aggressive_loop_done = true
              end

              expired_sample.map(&:first)
            end
            
            
            mutex.synchronize do
              # delete the key from the hash
              puts "Keys deleted: #{keys_to_delete.size}" unless keys_to_delete.empty?
              
              keys_to_delete.each do |key|
                expire_keys.delete(key)
                store.delete(key)
              end
            end
    
            elapsed_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
            # break the loop clean up task >= TIME_BUDGET
            if elapsed_time > TIME_BUDGET
              aggressive_loop_done = true
            end
          end
    
          execution_time = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
          # calcuate sleep time
          # max of 0 and cycle interval - total executed time
          sleep_time = [0, CYCLE_INTERVAL - execution_time].max
          sleep(sleep_time)
        end
      end
    end
  end
end

File: python/src/active_expire.py
import asyncio
import time
import random

class RedisActiveExpire:
  CYCLE_INTERVAL = 5
  TIME_BUDGET = 0.25 * CYCLE_INTERVAL

  @classmethod
  async def run(cls, store: dict, expire_store):
    while True:
      start_time: float = time.monotonic()
      aggressive_loop_done: bool = False

      while not aggressive_loop_done:
        # if expire_store is empty
        # break the inner loop
        if not expire_store:
          break

        # sample 20 pairs from the store
        sample_pairs: dict = {}
        if len(expire_store) > 20:
          sample_list: list = random.sample(sorted(expire_store), 20)
          sample_pairs = { k:v for k,v in expire_store.items() if k in sample_list }
        else:
          sample_pairs = expire_store
        
        # use current time in ms
        current_time: float = time.time() * 1000
        expire_keys: list = [k for k, v in sample_pairs.items() if current_time > v]

        # check if num of expire keys from sample pairs is greater than 25%
        expired_keys_above_threshold: bool = len(expire_keys) / len(sample_pairs) >= 0.25
        aggressive_loop_done = False if expired_keys_above_threshold else True

        # delete the keys from the store
        for key in expire_keys:
          print(f"Deleting {len(expire_keys)} expired keys")
          store.pop(key)
          expire_store.pop(key)
        
        # check the time budget
        elapsed_time: float = time.monotonic()
        executed_time: float = elapsed_time - start_time
        if executed_time > cls.TIME_BUDGET:
          aggressive_loop_done = True
      
      execution_time:float = time.monotonic()
      # get sleep time
      sleep_time: float = max([0, cls.CYCLE_INTERVAL - execution_time])
      
      await asyncio.sleep(sleep_time)
      
